### ISSUE TITLE
Cannot receive UDP traffic in Docker container in case of UDP server being located at Docker host

### ISSUE THREAD

**Issue by igorastds** [docker/docker#15127]
Description of problem:

When issuing UDP connections from docker container to docker host, no incoming UDP packets are received in container.
UDP works for accessing external resources.
TCP works for accessing docker host.

`docker version`:

Mostly tested on 

Client:
 Version:      1.8.0-dev
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   b900aaa
 Built:        Fri Jul 17 15:15:47 UTC 2015

Still actual in: 

 Go version:   go1.4.2
 Git commit:   9d3ad6d
 Built:        Wed Jul 29 16:26:04 UTC 2015

`docker info`:

Containers: 283
Images: 687
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.2███████
Operating System: Gentoo/Linux
CPUs: 8
Total Memory: 7.746 GiB
Name: ███████
ID: 7WYK:23Z2:J7KR:JFF2:QMLD:E5Z6:FF5O:LAAA:XOW4:56ZR:BOSD:WWAI

`uname -a`: Linux  ███████ 4.1.2-███████ #2 SMP Mon Jul 20 14:13:03 CEST 2015 x86_64 Intel(R) Core(TM) i7-4790 CPU @ 3.60GHz GenuineIntel GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.): physical

How reproducible:

Steps to Reproduce:
1. you need to start any UDP echo server (nmap's ncat works just fine: `ncat -e /bin/cat -k -u -l 5151`, also I rechecked with `udpqotd`, yeah that one from old Perl Cookbook) on Docker host outside of Docker (not tested starting server in docker container)
2. you need to test it with "netcat -u ip port" to see if you receive incoming traffic
3. you need to start "netcat -u ip port" in your favorite minimal-shell container and try accessing host, then verify your results by testing various TCP services and UDP echo services located outside of your docker host.

Actual Results:

- No incoming traffic received from UDP server when running in on docker host
- Incoming traffic successfully received from UDP server when running in on outside host

Expected Results:

- Incoming traffic received into docker from UDP server running in on docker host; it doesn't looks like a security problem because host is still accessible via TCP.

Additional info:

 - test container baseimage ubuntu:14.04 
 - firewalls are off, accessing TCP ports at docker host works
 - no --net or  flags given . Routing/traces seem to be ok. No --userland-proxy option is set in daemon config. 
 - no sysctl changes done; reboot was performed to verify results.

**Comment by chpio** [comment]
same here with docker v1.9.1 ("a34a1d5-dirty") @ Arch

:+1: please fix this bug

**Comment by chpio** [comment]
is there a workaround for this?

---

maybe it's an iptables bug. i have no other iptables rules, besides the docker rules, 2 networks (docker0 + 1 other). My host has the ip 172.16.0.1/16.

```
# iptables-save 
# Generated by iptables-save v1.4.21 on Sun Jan 17 10:28:50 2016
*nat
:PREROUTING ACCEPT [8794:1217326]
:INPUT ACCEPT [8773:1215606]
:OUTPUT ACCEPT [17899:2751833]
:POSTROUTING ACCEPT [17898:2752201]
:DOCKER - [0:0]
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
-A OUTPUT ! -d 127.0.0.0/8 -m addrtype --dst-type LOCAL -j DOCKER
-A POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE
-A POSTROUTING -s 172.19.0.0/16 ! -o br-bcfe92869fd5 -j MASQUERADE
COMMIT
# Completed on Sun Jan 17 10:28:50 2016
# Generated by iptables-save v1.4.21 on Sun Jan 17 10:28:50 2016
*filter
:INPUT ACCEPT [1364822:334302629]
:FORWARD ACCEPT [13:988]
:OUTPUT ACCEPT [1227632:7107850879]
:DOCKER - [0:0]
-A FORWARD -s 172.19.0.0/16 -d 172.17.0.0/16 -j DROP
-A FORWARD -s 172.17.0.0/16 -d 172.19.0.0/16 -j DROP
-A FORWARD -o docker0 -j DOCKER
-A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i docker0 ! -o docker0 -j ACCEPT
-A FORWARD -i docker0 -o docker0 -j ACCEPT
-A FORWARD -o br-bcfe92869fd5 -j DOCKER
-A FORWARD -o br-bcfe92869fd5 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i br-bcfe92869fd5 ! -o br-bcfe92869fd5 -j ACCEPT
-A FORWARD -i br-bcfe92869fd5 -o br-bcfe92869fd5 -j ACCEPT
COMMIT
# Completed on Sun Jan 17 10:28:50 2016
```

**Comment by LK4D4** [comment]
ping @mrjana @sanimej @aboch 
Could you pls try and close or fix issue?

**Comment by TimothyKlim** [comment]
Unbelievable, bug still in current docker version and no one cares from docker dev team (@thaJeztah).

**Comment by mavenugo** [comment]
@TimothyKlim sorry about that. this fell out of radar. We will take a look at it soon.

**Comment by TimothyKlim** [comment]
1 month later and still no progress

**Comment by chrisxaustin** [comment]
I'm having a [possibly] related issue. 

My app receives UDP syslog messages from most sources with the original source address preserved, but one device has the source rewritten to the docker0 interface's IP (seen by running tcpdump on docker0). 

I tried disabling the userland proxy as an experiment. The ones that were working before still are, but that one source never makes it to docker0 now, which is obviously concerning.

Ubuntu 16.04.2, docker 1.12.6, docker-compose 1.8.0, launched w/ docker-compose up

**Comment by cizixs** [comment]
The problem happens because host has multiple network interfaces(`eth0` and  `docker0` in this case), and when replying UDP, server choose the wrong IP as the source address.

Running `tcpdump` to capture packets on the wire gives:

```
14:42:11.861765 IP 172.17.0.3.37218 > 172.16.13.13.5151: UDP, length 6
14:42:11.863471 IP 172.17.0.1.5151 > 172.17.0.3.37218: UDP, length 6
14:42:11.863520 IP 172.17.0.3 > 172.17.0.1: ICMP 172.17.0.3 udp port 37218 unreachable, length 42
```
- The first packet means client sends a UDP request to server, with source address 172.17.0.3(container IP address), and destination address 172.16.13.13(host eth0 address)
- The second packets is the response, however, source address becomes `172.17.0.1`(docker0 ip address), instead of `172.16.13.13` as we expected
- The third packet is sent by client, telling server that the second packet is INVALID

As we can see the real cause is: **the second packet chooses the wrong source address!** And somehow, client does not recognize it, thinks it is INVALID, does not pass it to the application, but instead replies with a ICMP packet.

Choosing packet source address is kernel's work, for UDP, by default it will use the IP address of the interface the packet will be sent to. And as to why kernel drops the second packet, is because `netfilter` records valid UDP connection(since UDP is connectionless, it only means `netfilter` stores <sourcess IP, source Port, destination IP, destination Port>), and drops those that are invalid(not in connection table).

As for the first problem, socket can actually control which source address should be used with a combination of `sendmsg` and `IP_PKTINFO` socket option. See how powerDNS does it: https://blog.powerdns.com/2012/10/08/on-binding-datagram-udp-sockets-to-the-any-addresses/

In conclusion, this is not a problem caused by docker(although it does create a `docker0` which is the beginning of it) . And the solution comes in many ways:

- use TCP instead of UDP, because TCP is connected protocol, kernel will use the right address
- server should listen on a specific IP address
- Change server socket logic, please refer to: https://github.com/netoptimizer/network-testing/blob/master/src/udp_example02.c

**Comment by lagorsse** [comment]
I don't know if my issue is link or not but I can't send or receiver udp message in the docker container, works fine with tcp.

nc -l -u -p 8888
nc -u 127.0.0.1 8888

I am using
EXPOSE 8888/udp
or -p 8888:8888/udp

Netcat works fine with tcp so I don't get why it doesn't work over udp.
I am using osx sierra.

**Comment by REPTILEHAUS** [comment]
Also hitting this issue.. any fix for this ?

running osx Sierra too

**Comment by donovankim** [comment]
+1

**Comment by JoseFMP** [comment]
+1

**Comment by hello2mao** [comment]
+1

**Comment by ahirner** [comment]
+1

**Comment by weishuyin** [comment]
mark

**Comment by ingomueller-net** [comment]
I am running into this using `ntplib` in Python and OpenNTPD using [this docker image](https://github.com/cturra/docker-ntp).

>     * use TCP instead of UDP, because TCP is connected protocol, kernel will use the right address
>     * server should listen on a specific IP address
>     * Change server socket logic, please refer to: https://github.com/netoptimizer/network-testing/blob/master/src/udp_example02.c

I can't use TCP -- NTP runs over UDP. I don't want to change the server or client socket logic -- they are third party code and should be working fine. How should I make the server listen on a specific address then?

**Comment by eubnara** [comment]
+1

**Comment by makitsune** [comment]
+1

**Comment by HerryYT** [comment]
+1

**Comment by drgnomage** [comment]
+1

**Comment by zll001** [comment]
++1

**Comment by Hamsterman** [comment]
> The problem happens because host has multiple network interfaces(`eth0` and `docker0` in this case), and when replying UDP, server choose the wrong IP as the source address.
> 
> Running `tcpdump` to capture packets on the wire gives:
> 
> ```
> 14:42:11.861765 IP 172.17.0.3.37218 > 172.16.13.13.5151: UDP, length 6
> 14:42:11.863471 IP 172.17.0.1.5151 > 172.17.0.3.37218: UDP, length 6
> 14:42:11.863520 IP 172.17.0.3 > 172.17.0.1: ICMP 172.17.0.3 udp port 37218 unreachable, length 42
> ```
> 
> * The first packet means client sends a UDP request to server, with source address 172.17.0.3(container IP address), and destination address 172.16.13.13(host eth0 address)
> * The second packets is the response, however, source address becomes `172.17.0.1`(docker0 ip address), instead of `172.16.13.13` as we expected
> * The third packet is sent by client, telling server that the second packet is INVALID
> 
> As we can see the real cause is: **the second packet chooses the wrong source address!** And somehow, client does not recognize it, thinks it is INVALID, does not pass it to the application, but instead replies with a ICMP packet.
> 
> Choosing packet source address is kernel's work, for UDP, by default it will use the IP address of the interface the packet will be sent to. And as to why kernel drops the second packet, is because `netfilter` records valid UDP connection(since UDP is connectionless, it only means `netfilter` stores <sourcess IP, source Port, destination IP, destination Port>), and drops those that are invalid(not in connection table).
> 
> As for the first problem, socket can actually control which source address should be used with a combination of `sendmsg` and `IP_PKTINFO` socket option. See how powerDNS does it: https://blog.powerdns.com/2012/10/08/on-binding-datagram-udp-sockets-to-the-any-addresses/
> 
> In conclusion, this is not a problem caused by docker(although it does create a `docker0` which is the beginning of it) . And the solution comes in many ways:
> 
> * use TCP instead of UDP, because TCP is connected protocol, kernel will use the right address
> * server should listen on a specific IP address
> * Change server socket logic, please refer to: https://github.com/netoptimizer/network-testing/blob/master/src/udp_example02.c

Thank you! - I disabled the second network adapter on my host and my UDP issue went away!

**Comment by Rahix** [comment]
I think the actual issue lies in the proxy implementation, see moby/libnetwork#1729.

**Comment by JamesAwesome** [comment]
Just hit this same issue. enabling `use-vc` didn't work for me, but if i do `nc -vz <host-ip> 53` all dns then works. 

running dnsmasq on the host

resolv.conf looks like:

```
nameserver 127.0.0.11
options ndots:0
```

i don't really know why the `use-vc` option doesn't work for me, even if i change `nameserver` to the ipv4 of the host

**Comment by JamesAwesome** [comment]
Apparently my issue was I needed to bind dnsmasq to a specific IP address not 0.0.0.0 as per https://blog.powerdns.com/2012/10/08/on-binding-datagram-udp-sockets-to-the-any-addresses/

**Comment by developez** [comment]
I got an issue with UDP packets and a docker-container. In my case, I solved that in my docker-compose file add "/udp" in the ports section.

```

ports:
   "5223:5223/udp"

```

Source: https://docs.docker.com/config/containers/container-networking/

**Comment by victorli** [comment]
i got the same issue when connect the udp port of container, maybe docker forward the tcp connections by default, if want to connect udp port of container, u should add one forward rule to iptables's PREROUTING chain which under the type nat. 
eg:
 `iptables -t nat -A PREROUTING -p udp --dport 1234 -j DNAT --to container-ip:1234`

**Comment by andbuitra** [comment]
In my case, using Technitium DNS on host mode, changing the "DNS Server Local End Points" from 0.0.0.0:53 to [ip]:53 allowed the containers to get DNS from UDP.

**Comment by ben-xD** [comment]
> I got an issue with UDP packets and a docker-container. In my case, I solved that in my docker-compose file adding "/udp" in the ports section.
> 
> ```
> 
> ports:
>    "5223:5223/udp"
> ```
> 
> Source: [docs.docker.com/config/containers/container-networking](https://docs.docker.com/config/containers/container-networking/)

Are you sure you have the same issue as in this thread, or different UDP issue? Using `ports` like that will make docker bind to the port. This means other applications on the host won't be able to bind to that port? I'm getting "connection refused" errors. Remember, the title of the issue says "UDP server being located at Docker host".

I don't think there's a workaround (to UDP) to this bug yet.

**Comment by the-hotmann** [comment]
Same issue here:
I wanted to activate HTTP3 on nginx.
It works with `network_mode: host`, but not in `network_mode: bridge` (what it will be in, when using ports)

So for me a partial-fix/partial-workaround was adding the public IP to the port-mapping of UDP:

```
    ports:
      - "443:443/tcp"
      - "123.123.123.123:443:443/udp"
```

Now local requests with:

```
docker run -it --rm alpine/curl-http3:latest curl -v -k --http3-only https://###PUBLIC_IP###:443/
```

worked with HTTP3, but external requests from different servers did not work.

ping @mrjana @sanimej @aboch @mavenugo @thaJeztah
I hope someone from docker can inspect, reproduce and fix this. Seems like this issue exists since a long time. (a little over 9 years now)

Btw, for debugging reasons, here the info you might need:

`docker version`:
```
Client: Docker Engine - Community
 Version:           27.3.1
 API version:       1.47
 Go version:        go1.22.7
 Git commit:        ce12230
 Built:             Fri Sep 20 11:41:00 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          27.3.1
  API version:      1.47 (minimum version 1.24)
  Go version:       go1.22.7
  Git commit:       41ca978
  Built:            Fri Sep 20 11:41:00 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.22
  GitCommit:        7f7fdf5fed64eb6a7caf99b3e12efcf9d60e311c
 runc:
  Version:          1.1.14
  GitCommit:        v1.1.14-0-g2c9f560
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

`docker info`:
```
Client: Docker Engine - Community
 Version:    27.3.1
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.17.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.29.7
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 4
  Running: 4
  Paused: 0
  Stopped: 0
 Images: 6
 Server Version: 27.3.1
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7f7fdf5fed64eb6a7caf99b3e12efcf9d60e311c
 runc version: v1.1.14-0-g2c9f560
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.0-122-generic
 Operating System: Ubuntu 22.04.5 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 7.751GiB
 Name: misumi-comsrv-prod
 ID: 1950039b-fe7e-4251-a5f6-f33bc7f1d34a
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Default Address Pools:
   Base: 10.208.0.0/16, Size: 24

WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

(`bridge-nf-call-iptables` AFAIK should be disabled, otherwise iptables/nftables will controll it - if it is disabled docker controlls it)

**Comment by akerouanton** [comment]
@the-hotmann I can't reproduce with a simple `netcat` server. Could you try running this instead of your nginx container and report whether you can access port 5000 from remote hosts?

```
$ docker run --rm -it --network testnet -p 5000:5000/udp nicolaka/netshoot nc -ulkv 5000
```

**Comment by the-hotmann** [comment]
It does not work .. but I might should add: I also run iptables/nftables on my server. It is configured to lock everything down!

I run exactly these command:

```
docker network create testnet
iptables -A INPUT -p udp -s 0.0.0.0/0 --dport 5000 -j ACCEPT
docker run --rm -it --name netshoot --network testnet -p 5000:5000/udp nicolaka/netshoot nc -ulkv 5000
```

On my different server I then:

```
nc -vzu ###PUBLIC_IP### 5000
hostname.domain.tld [###PUBLIC_IP###] 5000 (?) open
```

But from the host itself it responds like this:

```
nc -vzu 10.208.2.2 5000
Connection to 10.208.2.2 5000 port [udp/*] succeeded!
```

even with the public IP from the host itself, it works:

```
nc -vzu ###PUBLIC_IP### 5000
Connection to ###PUBLIC_IP### 5000 port [udp/*] succeeded!
```

Does this help?

**Comment by akerouanton** [comment]
> I might should add: I also run iptables/nftables on my server.
> 
> docker run --rm -it --name netshoot --network testnet -p 5000:5000/udp nicolaka/netshoot nc -ulkv 5000

Note that this is not doing what you think it does, as per documented here: https://docs.docker.com/engine/install/ubuntu/#firewall-limitations.

> nc -vzu ###PUBLIC_IP### 5000
>
> but the docker does not print the "X" characters!

FWIW, I tend to prefer `echo foobar | nc ...`. The reason is that on Linux, `stdout` is _line-buffered_. That means, nothing will be printed by the listening netcat until it receives a line-break (eg. `echo ...`). See this [SO answer](https://stackoverflow.com/questions/3723795/is-stdout-line-buffered-unbuffered-or-indeterminate-by-default#answer-3746795).

Anyway, if you can't still reproduce the issue with `echo foobar | ...`, could you please run [iptables-tracer](https://github.com/akerouanton/iptables-tracer) on your docker host and post its output here? This would help us see if some iptables rule is broken.

```
$ wget -O iptables-tracer https://github.com/akerouanton/iptables-tracer/releases/download/v0.2/iptables-tracer-amd64
$ sudo mv iptables-tracer /usr/local/sbin/iptables-tracer
$ sudo chmod +x /usr/local/sbin/iptables-tracer
$ sudo iptables-tracer -family ipv4 -filter 'udp port 5000'
```

**Comment by the-hotmann** [comment]
Thanks for the hint .. I guess this was the hint.

I did use `tcpdump -i any port 5000 and udp` and noticed, that the company my server is located at has locked down `udp` for external IPs.

Anyway, I will ask them to open it and redo these tests:

1. Nginx HTTP3 Test (with `network_mode: bridge` & ports without explicit IP, from external server)
2. Nginx HTTP3 Test (with `network_mode: bridge` & ports with explicit IP, from external server)
3. Nginx HTTP3 Test (with `network_mode: host`, from external server)
4. netcat with udp port 5000

and report back if this really was the only problem, or if there are some more.
Then at least we know if this issue can be closed.


Thanks in advance!

**Comment by YodaDaCoda** [comment]
I encountered this issue running an Enshrouded game server, which communicates entirely over UDP, with Docker version 27.2.0, kernel 6.6.32.

The only thing that seemed to fix it was to include the local ip in the port mapping as has already been mentioned in this issue.
i.e.
```diff
  ports:
-  - "15637:15637/udp"
+  - "192.168.2.4:15637:15637/udp"
```

The issue was intermittent, with clients being able to connect for a few minutes after restarting the container, before new incoming connections stopped working (while existing connections continued to work).

This suggests the issue is in the kernel / caused by netfilter as suggested by @cizixs in https://github.com/moby/moby/issues/15127#issuecomment-323663787, though I haven't gone to the extent of using tcpdump to confirm the specifics of this behaviour.

I mention this mostly to raise that using netcat or similar from a single IP/port may not be sufficient to replicate the behaviour - using multiple hosts or at least multiple source ports may be required to properly trigger the behaviour.

**Comment by akerouanton** [comment]
@YodaDaCoda Thanks for highlighting @cizixs's comment -- I only took a quick look at this thread yesterday before answering.

@cizixs's issue seems slightly different than yours. IIUC, in their case, the UDP client is in a container, and the UDP server is running on the host. In your case, the UDP server is in a container, and UDP clients are remote hosts.

> As we can see the real cause is: the second packet chooses the wrong source address! 

That seems to be what https://blog.powerdns.com/2012/10/08/on-binding-datagram-udp-sockets-to-the-any-addresses describes indeed -- UDP server should use `setsockopt(IP_PKTINFO)` to avoid that.

But in your case @YodaDaCoda, it's hard to tell what's going on without a tcpdump. I feel like this has something to do with conntracking.

I think there's room for improvement for us here. Either in the form of a note in our docs, or by improving our iptables rules. I'll keep this issue on my todo list, and whenever I've some free time, I'll try to reproduce it by simulating a network with lots of nodes, and running multiple UDP clients on each.

It'd be great if you (or anyone else facing this issue) could provide a raw tcpdump, as well as the output of `conntrack -S` (taken at multiple intervals, once the issue appears). [pwru](https://github.com/cilium/pwru) might be interesting to run too, to see where the kernel drops datagrams (eg. `sudo pwru --output-tuple 'udp port 51011'`).

**Comment by jb0n** [comment]
@akerouanton I have a situation that might be helpful to your debugging efforts (hopefully we can help each other!). I've been having UDP/docker issues going ~2 years back. Missing packets sometimes. I usually am able to put a container in host mode and it's not a problem. I've been trying to stop running that way recently and was trying to debug issues when my colleague pointed me to this thread. Long story short I work on a network scanner and tracked down a major discrepancy with replies from certain protocols.

I took PCAPs and pwru outputs of scanning a /21 for SSDP specifically (UDP 1900) from the same host. I didn't include a conntrack (but I saw no dropped packets either). I have both the "host" captures and the "docker" captures, referring to the network. The PCAPs look OK to me actually, I see the bytes on the wire, but the program doesn't see all of them. My program consistently sees 46 IPs responding in host mode, 18 responding in docker networking mode. 

What it does is bind a single socket to the src IP, lets the OS pick an ephemeral port, then sends to all IPs in the /21 and reads off the same socket in a couple of goroutines. It will retry unless it gets a response.[docker-udp.tar.gz](https://github.com/user-attachments/files/17426161/docker-udp.tar.gz)

Let me know if there is anything else I can do to assist debugging this issue. Thanks.

**Comment by the-hotmann** [comment]
> * Nginx HTTP3 Test (with `network_mode: bridge` & ports without explicit IP, from external server)
> * Nginx HTTP3 Test (with `network_mode: bridge` & ports with explicit IP, from external server)
> * Nginx HTTP3 Test (with `network_mode: host`, from external server)

Comming back to my promisse :)

1. After configuring the FW and using bridge mode: I COULD CONNECT with HTTP/3 (I just have one interface.. so maybe thats why)
2. After configuring the FW and using bridge mode with explicit IP: I COULD CONNECT with HTTP/3
3. after configuring the FW and using host mode: I COULD CONNECT with HTTP/3

But keep in mind, that when I used bridge mode I connected to the backend, container vis its docker `hostname:port`.
When using host mode I connected to is using `127.0.0.1:port`.

I skipped test 4 as it now is irrelevant.
For me personally this issue is solved,

**Comment by pasbec** [comment]
Hi!

Has there been any investigation on the UDP regression since Docker 5.28? I still have to pin my Docker server to docker 5.27.5 to get UDP working properly in bridged mode for [Lyrion Music Server (LMS)](https://hub.docker.com/r/lmscommunity/lyrionmusicserver) without any firewall.

Even GitHub Copilot tells me that UDP/multicast is broken in Docker 5.28+:

<img width="796" height="724" alt="Image" src="https://github.com/user-attachments/assets/79ac26de-480a-4ea5-ae0a-af9d78849616" />

**Edit:**

Seems like this partly related to raw iptable rules which did not exist before in Docker 27. In Docker 28+, I can see someting like:
```bash
sudo iptables -t raw -L -n -v
```
```text
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 DROP       0    --  !br-88d4e2c7318c *       0.0.0.0/0            172.25.0.2
...
```

From #49832 I found an option which, can be applied e.g. via compose:
```compose
networks:
  lms:
    driver: bridge
    driver_opts:
      com.docker.network.bridge.trusted_host_interfaces: eth0
```

This turns the breaking raw iptables stuff into something which works:
```text
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     0    --  eth0   *       0.0.0.0/0            172.25.0.2
    0     0 DROP       0    --  !br-88d4e2c7318c *       0.0.0.0/0            172.25.0.2
...
```

But it is still not working correctly.
